feat: Add WNFS plugin

icidasset opened the pull request:
### Motivation / Background

Adds a WNFS plugin that allows for private image uploads in Composer which are synced to other devices using the Blob (edge) service.

### Details

Initial working version of the idea.
Connects to the `http://localhost:8787` blob service, is there a way to configure the URL for this?
Depends on https://github.com/dxos/edge/pull/91

Demo:
https://discord.com/channels/837138313172353095/1142114470479536368/1284153524103549009

Needs some fine tuning, notes/todos:
- Codemirror plugin needs improving (I don't have any experience with this, just threw something together)
- A WNFS is built up out of several merkle DAGs, it makes an HTTP request for each one of them (only 1st time, after = cache). Both for retrieval and storage. So this might turn out slow for big files, could write a batching mechanism.
- Related to the above. What if the person closes the browser during the remote blob storage call? The blob service won't have the data. I guess we need some guard rails here. Like tracking if the blob upload was successful, if not, retry on next app launch. Or maybe, prevent the browser from closing during upload.
- Still needs to check if the browser is online or not. If not, don't try to push to the remote blob store and track which ones need to uploaded later on when the connection comes back.
- Loading animations for images.

icidasset commented on packages/sdk/config/src/plugin/definitions.ts:
```diff
@@ -25,8 +25,8 @@ export const definitions = ({
-    __CONFIG_DEFAULTS__: configPath ?? resolve(CWD, 'dx.yml'),
-    __CONFIG_ENVS__: envPath ?? resolve(CWD, 'dx-env.yml'),
+    __CONFIG_DEFAULTS__: configPath && configPath.length ? configPath : resolve(CWD, 'dx.yml'),
+    __CONFIG_ENVS__: envPath && envPath.length ? envPath : resolve(CWD, 'dx-env.yml'),
```
Environment wouldn't load for me without this.

wittjosiah replied:
interesting... wonder why that's not come up before. I think it could be shortened to `configPath?.length` but seems reasonable to me

icidasset replied:
Hmm.. I'm guessing it's because I'm using this shell that almost no one else uses (elvish).

icidasset commented on packages/ui/react-ui-editor/src/extensions/markdown/image.ts:
```diff
@@ -63,6 +63,9 @@ const buildDecorations = (from: number, to: number, state: EditorState) => {
         if (urlNode) {
           const hide = state.readOnly || cursor < node.from || cursor > node.to;
           const url = state.sliceDoc(urlNode.from, urlNode.to);
+          if (url.match(/^https?:\/\//) === null) {
```
Avoid rendering non-HTTP urls to images.

wittjosiah replied:
Could you leave a comment here to explain why this is needed?

wittjosiah commented on packages/plugins/experimental/plugin-wnfs/src/load.ts:
```diff
@@ -0,0 +1,61 @@
+
+import { Rng, store } from './common';
+
+// 🚀
```
prefer descriptive comments, I'm not sure this communicates intent clearly enough

wittjosiah commented on packages/plugins/experimental/plugin-wnfs/src/meta.tsx:
```diff
@@ -0,0 +1,18 @@
+  name: 'WNFS',
+  description: 'Upload & view files with WNFS.',
+  tags: ['experimental'],
+  iconComponent: (props: IconProps) => <FileCloud {...props} />,
```
if you pull from `main` these were recently updated to be icon strings which index a sprite sheet. This will need to align with that for the icon to work now

wittjosiah reviewed (commented):
looking good, nice work!

one thing I couldn't find was where the `FileType` objects were being created. I was looking because I was thinking it would probably be good to include some sort of meta key on them to indicate that they were associated with the blob store service

icidasset commented on packages/plugins/experimental/plugin-wnfs/src/components/index.ts:
```diff
@@ -0,0 +1,11 @@
+
+import React from 'react';
+
+// Lazily load components for content surfaces.
```
Yeah, that's where I started months ago.
I have no idea how those components work 😅 
Let me know what to do there and I'll make another PR for that.

richburdon replied:
this is fine.

icidasset commented:
Thanks @wittjosiah! I've applied your feedback. Also not sure where the actual `FileType` objects are created, I vaguely remember something about the uploader interface of composer/markdown-plugin? 🤷 Definitely not in this new plugin though.

icidasset commented on packages/plugins/experimental/plugin-wnfs/src/blockstore.ts:
```diff
@@ -0,0 +1,82 @@
+  constructor() {
+    super();
+
+    this.#apiHost = 'http://localhost:8787';
```
@wittjosiah This is supposed to target the Blob service, is there a production instance yet?

icidasset replied:
@richburdon Any thoughts?

richburdon replied:
obtain from the config object (useConfig).
add `runtime.services.blobstore.endpoint`

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/WnfsPlugin.tsx:
```diff
@@ -0,0 +1,118 @@
+          );
+
+          const [_, updatedForest] = await result.rootDir.store(result.forest, wnfsStore, new Rng());
+
```
i think group these lines (no blank lines) since they are closely related. perhaps adding a comment to orient people unfamiliar with wnfs.

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/common.ts:
```diff
@@ -0,0 +1,81 @@
+    await blockstore.put(decodedCid, bytes);
+  };
+
+  // Don't hash blocks with the rs-wnfs default Blake 3, sha256 has better support
```
our comments standard is to put a period at the end of the sentence (to make sure it isn't truncated).

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/common.ts:
```diff
@@ -0,0 +1,81 @@
+import * as crypto from '@dxos/crypto';
+import { type Space } from '@dxos/react-client/echo';
+
+// PATH
```
nit

```
//
// Path
//
```

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FileMain.tsx:
```diff
@@ -0,0 +1,23 @@
+import { type FileType } from '../types';
+
+const FileMain: FC<{ file: FileType }> = ({ file }) => {
+  const url = 'TODO';
```
?

icidasset replied:
I haven't worked on these components, was planning to do that in a separate PR. I'd also like some input on how to proceed with these components.

icidasset replied:
@richburdon See the comment above.

richburdon replied:
ok we're doing a rewrite of mosaic, so i think fine to leave this.

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FilePreview.tsx:
```diff
@@ -0,0 +1,24 @@
+/**
+ * File/content preview iframe.
+ */
+export const FilePreview: FC<FilePreviewProps> = ({ type, url, className }) => {
```
you might be copying one of our antipatterns, since we've done this in the past, but now prefer

FilePreview = ({ ... }: FilePreviewProps) =>

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FileSlide.tsx:
```diff
@@ -0,0 +1,20 @@
+import { type FileType } from '../types';
+
+const FileSlide: FC<{ file: FileType; cover?: boolean }> = ({ file, cover }) => {
+  const url = 'TODO';
```
?

icidasset replied:
Same comment as https://github.com/dxos/dxos/pull/6636#discussion_r1796784021

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/extensions/image.ts:
```diff
@@ -0,0 +1,247 @@
+      if (eff === undefined) {
+        return value;
+      }
+      return value.map(tr.changes).update({
```
i think this would be clearly (and more compact) if you just put this inline inside the for/if loop.

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/extensions/image.ts:
```diff
@@ -0,0 +1,247 @@
+      }
+
+      const { from, to } = changedRange(update);
+
```
remove blank line (group statements by context)

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/extensions/image.ts:
```diff
@@ -0,0 +1,247 @@
+          });
+        })
+        .catch((err) => {
+          throw new Error(err);
```
log.catch(err)?

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FileMain.tsx:
```diff
@@ -0,0 +1,23 @@
+//
```
is it possible to create a unit test or a react story to demo upload and preview of images? e.g., look at TextEdtitor.stories.tsx (low-level) or mermaid.stories.tsx (which is another plugin).

icidasset replied:
Sure yeah, I'll add that to this PR. I assume without involving the blob service right?

icidasset replied:
@richburdon Added one.
![Screenshot 2024-10-24 at 19 03 01](https://github.com/user-attachments/assets/780474f4-257a-4eb3-810a-39d9827453c4)

Renders the uploaded images.

richburdon reviewed (commented):
Really nice!

icidasset commented:
Rebased, hopefully interpreted the changes on `main` correctly.
Other changes:
- Added story component.
- Used a CID as the file name of the uploaded file, in case multiple images with the same name are uploaded (but with different contents).

wittjosiah commented:
> I got an error of how I used useConfig in the plugin, so I used parseClientPlugin instead. Is that the right approach @wittjosiah ?

Yeah that looks right. `useConfig` is a react hook and `ready` runs outside of the realm of react so that would be why it wouldn't work.

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/common.ts:
```diff
@@ -0,0 +1,85 @@
+// STORE
+//
+
+export const store = (blockstore: Blockstore): WnfsBlockStore => {
```
why not make this a class?

icidasset replied:
Could do that, but thought a function was more appropriate seeing how I use the resulting `WnfsBlockStore`. Pattern goes like so:
```ts
const blockstore: Blockstore = makeIndexedDbBlockstore()

await wnfsInstance.write(
  ...,
  common.store(blockstore)
  // Feels a bit nicer to write, as opposed to:
  new common.Store(blockstore)
  // And using a class might give people more the idea you're creating a new store (which isn't the case)
)
```

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FileCard.tsx:
```diff
@@ -0,0 +1,45 @@
+  },
+);
+
+export default FileCard;
```
in general we prefer explicit `export const`

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/components/FilePreview.tsx:
```diff
@@ -0,0 +1,26 @@
+export type FilePreviewProps = {
+  type: string;
+  url: string;
+  className?: string;
```
`type FilePreviewProps = ThemedClassName<{ type, url }>`

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/extensions/image.stories.tsx:
```diff
@@ -0,0 +1,138 @@
+};
+
+export default {
+  title: 'plugin-wnfs/image',
```
this has changed recently: 'pluings/plugin-wnfs/image'

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/load.ts:
```diff
@@ -0,0 +1,63 @@
+
+// LOAD
+// ----
+
```
```
//
// Load
//
```

(as above)

richburdon commented on packages/plugins/experimental/plugin-wnfs/src/load.ts:
```diff
@@ -0,0 +1,63 @@
+
+  const rng = new Rng();
+  const forest = new PrivateForest(rng);
+
```
remove black links to logically group related statements?

richburdon reviewed (approved):
Great work steven!
